src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1455,6 +1533,44 @@ public ValidateSetAttribute(params string[] validValues)
 
             _validValues = validValues;
         }
+
+        /// <summary>
+        /// Initializes a new instance of the ValidateSetAttribute class.
+        /// Valid values is returned dynamically from a custom class implementing 'IValidateSetValuesGenerator' interface.
+        /// </summary>
+        /// <param name="valuesGeneratorType">class that implements the 'IValidateSetValuesGenerator' interface</param>
+        /// <exception cref="ArgumentException">for null arguments</exception>
+        public ValidateSetAttribute(Type valuesGeneratorType)
+        {
+            if (!typeof(IValidateSetValuesGenerator).IsAssignableFrom(valuesGeneratorType))
```

Comment:
And I think when validation fails, the thrown exception should be informational. `NewArgumentException("valuesGeneratorType")` doesn't indicate what is the problem.